File: src/main/java/name/gyger/jmoney/report/ReportController.java
package name.gyger.jmoney.report;

import name.gyger.jmoney.account.Entry;
import name.gyger.jmoney.account.EntryService;
import name.gyger.jmoney.util.DateUtil;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.Date;
import java.util.List;

@RestController
@RequestMapping("/rest/reports")
public class ReportController {

    private final ReportService reportService;

    public ReportController(ReportService reportService, EntryService entryService) {
        this.reportService = reportService;
    }


    @RequestMapping(path = "/balances", method = RequestMethod.GET)
    public List<Balance> getBalances(@RequestParam(value = "date", required = false) String dateString) {
        Date date = DateUtil.parse(dateString);
        return reportService.getBalances(date);
    }

    @RequestMapping(path = "/cash-flows", method = RequestMethod.GET)
    public List<CashFlow> getCashFlow(@RequestParam(value = "fromDate", required = false) String fromDateString,
                                      @RequestParam(value = "toDate", required = false) String toDateString) {
        Date fromDate = DateUtil.parse(fromDateString);
        Date toDate = DateUtil.parse(toDateString);

        return reportService.getCashFlow(fromDate, toDate);
    }

    @RequestMapping(path = "/entries-with-category", method = RequestMethod.GET)
    public List<Entry> getEntries(@RequestParam(value = "categoryId") long categoryId,
                                  @RequestParam(value = "fromDate") String fromDateString,
                                  @RequestParam(value = "toDate") String toDateString) {
        Date from = DateUtil.parse(fromDateString);
        Date to = DateUtil.parse(toDateString);
        return reportService.getEntriesForCategory(categoryId, from, to);
    }

    @RequestMapping(path = "/consitency/inconsistent-split-entries", method = RequestMethod.GET)
    public List<Entry> getInconsistentSplitEntries() {
        return reportService.getInconsistentSplitEntries();
    }

    @RequestMapping(path = "/consitency/entries-without-category", method = RequestMethod.GET)
    public List<Entry> getEntriesWithoutCategory() {
        return reportService.getEntriesWithoutCategory();
    }

}

File: src/main/kotlin/name/gyger/jmoney/report/ReportController.kt
package name.gyger.jmoney.report

import name.gyger.jmoney.account.Entry
import name.gyger.jmoney.account.EntryService
import name.gyger.jmoney.util.parse
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController


@RestController
@RequestMapping("/api/reports")
class ReportController(private val reportService: ReportService,
                       private val entryService: EntryService) {

    @GetMapping("/balances")
    fun getBalances(@RequestParam(value = "date", required = false) dateString: String?): List<Balance> {
        val date = parse(dateString)
        return reportService.getBalances(date)
    }

    @GetMapping("/cash-flows")
    fun getCashFlow(@RequestParam(value = "fromDate") fromDateString: String,
                    @RequestParam(value = "toDate") toDateString: String): List<CashFlow> {
        val fromDate = parse(fromDateString)
        val toDate = parse(toDateString)
        return reportService.getCashFlow(fromDate, toDate)
    }

    @GetMapping("/entries-with-category")
    fun getEntries(@RequestParam(value = "categoryId") categoryId: Long,
                   @RequestParam(value = "fromDate") fromDateString: String,
                   @RequestParam(value = "toDate") toDateString: String): List<Entry> {
        val from = parse(fromDateString)
        val to = parse(toDateString)
        return entryService.getEntriesForCategory(categoryId, from, to)
    }

    @GetMapping("/consistency/inconsistent-split-entries")
    fun getInconsistentSplitEntries(): List<Entry> {
        return reportService.getInconsistentSplitEntries()
    }

    @GetMapping("/consistency/entries-without-category")
    fun getEntriesWithoutCategory(): List<Entry> {
        return entryService.getEntriesWithoutCategory()
    }

}
